File: server/protocol.py
"""
Wire protocol for the intercom system.

Every UDP packet starts with a fixed header:

    byte 0        : message type (see MsgType below)
    bytes 1-4     : session token (uint32, big-endian). 0 for pre-auth packets.
    bytes 5+      : type-specific payload

This is intentionally simple and stateless-ish: the session token in every
packet is what lets the server figure out which phone sent it without a
persistent TCP connection. If a phone drops off wifi and comes back, it just
resumes sending packets with the same token and everything picks back up
with no reconnect handshake needed.
"""

import struct

HEADER_FMT = "!BI"          # type (1 byte), token (4 bytes unsigned)
HEADER_SIZE = struct.calcsize(HEADER_FMT)


class MsgType:
    AUTH_REQUEST = 0x01     # client -> server: payload = UTF-8 password + "\x00" + UTF-8 device name
    AUTH_RESPONSE = 0x02    # server -> client: payload = 1 byte status (0=fail,1=ok) + 4 byte token if ok
    HEARTBEAT = 0x03        # client -> server: payload = empty
    HEARTBEAT_ACK = 0x04    # server -> client: payload = empty
    AUDIO_FRAME = 0x05      # both directions: payload = 4 byte seq + 8 byte origin timestamp (ms) + PCM audio
    LOGOUT = 0x06           # client -> server: payload = empty
    PING = 0x07             # client -> server: payload = 8 byte timestamp (ms), echoed back immediately
    PONG = 0x08             # server -> client: payload = same 8 byte timestamp, unmodified
    LOOPBACK_ON = 0x09      # client -> server: payload = empty. Server stops excluding this client's own
                            # audio from its mix, for latency self-testing.
    LOOPBACK_OFF = 0x0A     # client -> server: payload = empty. Restores normal self-exclusion.
    MUTE_ON = 0x0B          # server -> client: payload = empty. Admin has force-muted this phone's mic.
    MUTE_OFF = 0x0C         # server -> client: payload = empty. Admin has lifted the forced mute.
    TEXT_MESSAGE = 0x0D     # server -> client: payload = UTF-8 message text, shown/rung on the phone


def pack_header(msg_type: int, token: int) -> bytes:
    return struct.pack(HEADER_FMT, msg_type, token)


def unpack_header(data: bytes):
    if len(data) < HEADER_SIZE:
        return None
    msg_type, token = struct.unpack(HEADER_FMT, data[:HEADER_SIZE])
    return msg_type, token, data[HEADER_SIZE:]


def make_auth_request(password: str, device_name: str = "") -> bytes:
    payload = password.encode("utf-8") + b"\x00" + device_name.encode("utf-8")
    return pack_header(MsgType.AUTH_REQUEST, 0) + payload


def parse_auth_request_payload(payload: bytes):
    """Returns (password, device_name) from an AUTH_REQUEST payload."""
    if b"\x00" in payload:
        password_bytes, name_bytes = payload.split(b"\x00", 1)
    else:
        password_bytes, name_bytes = payload, b""
    try:
        return password_bytes.decode("utf-8"), name_bytes.decode("utf-8")
    except UnicodeDecodeError:
        return "", ""


def make_auth_response(ok: bool, token: int = 0) -> bytes:
    payload = struct.pack("!B", 1 if ok else 0)
    if ok:
        payload += struct.pack("!I", token)
    return pack_header(MsgType.AUTH_RESPONSE, 0) + payload


def make_heartbeat(token: int) -> bytes:
    return pack_header(MsgType.HEARTBEAT, token)


def make_heartbeat_ack(token: int) -> bytes:
    return pack_header(MsgType.HEARTBEAT_ACK, token)


def make_audio_frame(token: int, seq: int, origin_timestamp_ms: int, pcm_bytes: bytes) -> bytes:
    return (
        pack_header(MsgType.AUDIO_FRAME, token)
        + struct.pack("!IQ", seq, origin_timestamp_ms)
        + pcm_bytes
    )


def make_logout(token: int) -> bytes:
    return pack_header(MsgType.LOGOUT, token)


def make_ping(token: int, timestamp_ms: int) -> bytes:
    return pack_header(MsgType.PING, token) + struct.pack("!Q", timestamp_ms)


def make_pong(token: int, timestamp_ms: int) -> bytes:
    return pack_header(MsgType.PONG, token) + struct.pack("!Q", timestamp_ms)


def make_loopback_on(token: int) -> bytes:
    return pack_header(MsgType.LOOPBACK_ON, token)


def make_loopback_off(token: int) -> bytes:
    return pack_header(MsgType.LOOPBACK_OFF, token)


def make_mute_on(token: int) -> bytes:
    return pack_header(MsgType.MUTE_ON, token)


def make_mute_off(token: int) -> bytes:
    return pack_header(MsgType.MUTE_OFF, token)


def make_text_message(token: int, text: str) -> bytes:
    return pack_header(MsgType.TEXT_MESSAGE, token) + text.encode("utf-8")


def parse_audio_frame_payload(payload: bytes):
    """Returns (seq, origin_timestamp_ms, pcm_bytes) from an AUDIO_FRAME payload (post-header)."""
    seq, origin_timestamp_ms = struct.unpack("!IQ", payload[:12])
    return seq, origin_timestamp_ms, payload[12:]


def parse_ping_payload(payload: bytes) -> int:
    return struct.unpack("!Q", payload[:8])[0]

File: android/app/src/main/java/com/fieldintercom/app/Protocol.kt
package com.fieldintercom.app

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

/**
 * Mirrors server/protocol.py exactly. If you change one, change the other.
 *
 * Header layout (5 bytes): [type: 1 byte][token: 4 bytes big-endian] then payload.
 */
object Protocol {
    const val AUTH_REQUEST: Byte = 0x01
    const val AUTH_RESPONSE: Byte = 0x02
    const val HEARTBEAT: Byte = 0x03
    const val HEARTBEAT_ACK: Byte = 0x04
    const val AUDIO_FRAME: Byte = 0x05
    const val LOGOUT: Byte = 0x06
    const val PING: Byte = 0x07
    const val PONG: Byte = 0x08
    const val LOOPBACK_ON: Byte = 0x09
    const val LOOPBACK_OFF: Byte = 0x0A
    const val MUTE_ON: Byte = 0x0B
    const val MUTE_OFF: Byte = 0x0C
    const val TEXT_MESSAGE: Byte = 0x0D

    const val HEADER_SIZE = 5

    // Must match server/config.py
    const val SAMPLE_RATE = 16000
    const val FRAME_MS = 20
    const val SAMPLES_PER_FRAME = SAMPLE_RATE * FRAME_MS / 1000 // 320
    const val FRAME_BYTES = SAMPLES_PER_FRAME * 2 // 16-bit PCM

    data class Header(val type: Byte, val token: Int, val payloadOffset: Int)

    fun parseHeader(data: ByteArray, length: Int): Header? {
        if (length < HEADER_SIZE) return null
        val buf = ByteBuffer.wrap(data, 0, HEADER_SIZE)
        val type = buf.get()
        val token = buf.int
        return Header(type, token, HEADER_SIZE)
    }

    fun makeAuthRequest(password: String, deviceName: String): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(AUTH_REQUEST.toInt())
        out.write(intToBytes(0))
        out.write(password.toByteArray(StandardCharsets.UTF_8))
        out.write(0) // null separator
        out.write(deviceName.toByteArray(StandardCharsets.UTF_8))
        return out.toByteArray()
    }

    fun makeHeartbeat(token: Int): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(HEARTBEAT.toInt())
        out.write(intToBytes(token))
        return out.toByteArray()
    }

    /** originTimestampMs: sender's local clock at capture time (elapsedRealtime-based). */
    fun makeAudioFrame(token: Int, seq: Int, originTimestampMs: Long, pcm: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(AUDIO_FRAME.toInt())
        out.write(intToBytes(token))
        out.write(intToBytes(seq))
        out.write(longToBytes(originTimestampMs))
        out.write(pcm)
        return out.toByteArray()
    }

    fun makeLogout(token: Int): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(LOGOUT.toInt())
        out.write(intToBytes(token))
        return out.toByteArray()
    }

    fun makePing(token: Int, timestampMs: Long): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(PING.toInt())
        out.write(intToBytes(token))
        out.write(longToBytes(timestampMs))
        return out.toByteArray()
    }

    fun makeLoopbackOn(token: Int): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(LOOPBACK_ON.toInt())
        out.write(intToBytes(token))
        return out.toByteArray()
    }

    fun makeLoopbackOff(token: Int): ByteArray {
        val out = ByteArrayOutputStream()
        out.write(LOOPBACK_OFF.toInt())
        out.write(intToBytes(token))
        return out.toByteArray()
    }

    /** Parses an AUTH_RESPONSE payload (bytes after the 5-byte header). Returns token, or null on failure. */
    fun parseAuthResponse(data: ByteArray, offset: Int, length: Int): Int? {
        if (length - offset < 1) return null
        val ok = data[offset].toInt() != 0
        if (!ok) return null
        if (length - offset < 5) return null
        return ByteBuffer.wrap(data, offset + 1, 4).int
    }

    data class AudioFramePayload(val seq: Int, val originTimestampMs: Long, val pcm: ByteArray)

    /** Parses an AUDIO_FRAME payload (bytes after the 5-byte header). */
    fun parseAudioFramePayload(data: ByteArray, offset: Int, length: Int): AudioFramePayload {
        val seq = ByteBuffer.wrap(data, offset, 4).int
        val originTs = ByteBuffer.wrap(data, offset + 4, 8).long
        val pcmStart = offset + 12
        val pcmLen = length - pcmStart
        return AudioFramePayload(seq, originTs, data.copyOfRange(pcmStart, pcmStart + pcmLen))
    }

    /** Parses a PONG payload (bytes after the 5-byte header). Returns the echoed timestamp. */
    fun parsePongPayload(data: ByteArray, offset: Int): Long {
        return ByteBuffer.wrap(data, offset, 8).long
    }

    /** Parses a TEXT_MESSAGE payload (bytes after the 5-byte header) as UTF-8 text. */
    fun parseTextMessagePayload(data: ByteArray, offset: Int, length: Int): String {
        return String(data, offset, length - offset, StandardCharsets.UTF_8)
    }

    private fun intToBytes(value: Int): ByteArray {
        return ByteBuffer.allocate(4).putInt(value).array()
    }

    private fun longToBytes(value: Long): ByteArray {
        return ByteBuffer.allocate(8).putLong(value).array()
    }
}
